### PR TITLE
Problem of username selection: improvements

### LINKED ISSUE #915: Problem of username selection: improvements
Right now we give user exactly the same name/id, that he has in the initial system (codehub/github). 
We need to support two main things on the frontend:
1) "Registration"-like page, where user will be able to change/update his nickname OR use one we have suggested;
2) Possibility to change nick-name (if it is not yet used by other people).

### PR DESCRIPTION
**Pull request by Cheshiriks**
* added Registration view
Closes #915

### CODE REVIEW

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt [hunk 1]
```diff
@@ -128,7 +128,7 @@ class WebSecurityConfig(
             it.authenticationSuccessHandler(
                 DelegatingServerAuthenticationSuccessHandler(
                     StoringServerAuthenticationSuccessHandler(configurationProperties),
-                    RedirectServerAuthenticationSuccessHandler("/#/projects"),
+                    RedirectServerAuthenticationSuccessHandler("/#/registration"),
```
Comment (sanyavertolet) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt:131]: Will it work only when it is the first time that the user is authenticated with his GitHub account?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+            }
+
+    /* companion object :
+        RStatics<RegistrationProps, RegistrationViewState, RegistrationView, Context<RequestStatusContext>>(
+            RegistrationView::class
+        ) {
+        init {
+            RegistrationView.contextType = requestStatusContext
+        }
+    }*/
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:266]: Why is this companion object commented?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    @Suppress(
+        "TOO_LONG_FUNCTION",
+        "EMPTY_BLOCK_STRUCTURE_ERROR",
+        "LongMethod",
+        "MAGIC_NUMBER",
+        "PARAMETER_NAME_IN_OUTER_LAMBDA",
+    )
+    override fun ChildrenBuilder.render() {
+        if (state.userInfo?.isActive == false) {
+            ReactHTML.main {
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:151]: Remove `ReactHTML` please

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+                                            }
+                                                ?: run {
+                                                    "img/undraw_profile.svg"
+                                                }
+                                            height = 260.0
+                                            width = 260.0
+                                        }
+                                    }
+                                    ReactHTML.form {
+                                        className = ClassName("needs-validation")
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:193]: Not sure if we need this because it should already be inside `inputTextFormRequired`

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    var isValidUserName: Boolean?
+}
+
+/**
+ * A Component for registration view
+ */
+@JsExport
+@OptIn(ExperimentalJsExport::class)
+class RegistrationView : AbstractView<RegistrationProps, RegistrationViewState>() {
+    private val fieldsMap: MutableMap<InputTypes, String> = mutableMapOf()
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:84]: I strongly recommend you to use state here instead of private field of class as far as it is more react style (afaik)

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/domain/UserSaveStatus.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    /**
+     * New user
+     */
+    NEW("User saved successfully"),
+
+    /**
+     * Update user
+     */
+    UPDATE("User update successfully"),
```
Comment (sanyavertolet) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/domain/UserSaveStatus.kt:24]:
```suggestion
    UPDATE("User successfully updated"),
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:266]: It's strange that the Diktat does not swear at this

Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:266]: Created issue https://github.com/saveourtool/diktat/issues/1510

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt (hunk 1 above)
Comment (Cheshiriks) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt:131]: Yes. We check if the user is registered. If so, move it to _projects_ page

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Why did you completely remove companion object? Don't we need it here? I thought you might want to process responses using `responseHandlerWithValidation`.

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/UsersDetailsController.kt [hunk 7]
```diff
@@ -44,33 +50,53 @@ class UsersDetailsController(
     @GetMapping("/{userName}")
     @PreAuthorize("permitAll()")
     fun findByName(@PathVariable userName: String): Mono<UserInfo> =
-            userRepository.findByName(userName).toMonoOrNotFound().map { it.toUserInfo() }
+            userRepository.findByName(userName)?.let { user ->
+                user.toMonoOrNotFound().map { it.toUserInfo() }
+            } ?: run {
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/UsersDetailsController.kt:55]:
```suggestion
            userRepository.findByName(userName)
                ?.toMonoOrNotFound()?.map { it.toUserInfo() }
                ?:
```

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt [hunk 8]
```diff
@@ -44,11 +47,12 @@ class UsersController(
...
     fun saveNewUser(@RequestBody user: User) {
         val userName = requireNotNull(user.name) { "Provided user $user doesn't have a name" }
-        userRepository.findByName(userName)?.let {
+        originalLoginRepository.findByName(userName)?.user?.let {
             logger.debug("User $userName is already present in the DB")
         } ?: run {
             logger.info("Saving user $userName to the DB")
-            userRepository.save(user)
+            val newUser = userRepository.save(user)
+            originalLoginRepository.save(OriginalLogin(user.name, newUser, user.source))
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt:55]: This should be encapsulated in `UserDetailsService`

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt [hunk 9]
```diff
@@ -44,11 +47,12 @@ class UsersController(
     @PostMapping("/new")
     fun saveNewUser(@RequestBody user: User) {
         val userName = requireNotNull(user.name) { "Provided user $user doesn't have a name" }
-        userRepository.findByName(userName)?.let {
+        originalLoginRepository.findByName(userName)?.user?.let {
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt:50]: Why are you checking `originalLoginRepository` here? What if we need to save a user with a name that exists in original logins but from a new authorization source?

Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt:50]: Why are you checking `originalLoginRepository` here? What if we need to save a user with a name that exists in original logins but from a new authorization source?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ValidateRepository.kt [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    @Query("""select if (count(*) = 0, true, false) from save_cloud.high_level_names where name = :name""", nativeQuery = true)
+    fun validateName(@Param("name") name: String): Long
+
+    /**
+     * @param name
+     */
+    @Query("""insert into save_cloud.high_level_names set name = :name""", nativeQuery = true)
+    @Modifying
+    fun saveHighName(@Param("name") name: String)
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ValidateRepository.kt:23]:
```suggestion
    fun saveHighLevelName(@Param("name") name: String)
```

File: save-backend/src/test/kotlin/com/saveourtool/save/backend/controller/OrganizationControllerTest.kt [hunk 11]
```diff
@@ -48,7 +48,7 @@ import java.util.*
     ProjectPermissionEvaluator::class,
     LnkUserProjectService::class,
     UserDetailsService::class,
-
+    OriginalLoginRepository::class,
```
Comment (petertrr) [inline save-backend/src/test/kotlin/com/saveourtool/save/backend/controller/OrganizationControllerTest.kt:51]: Should be either as `@Import` or `@MockBean`

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt [hunk 12]
```diff
@@ -19,10 +21,17 @@ import kotlinx.serialization.Serializable
...
+ * @property isActive
+ * @property oldName
+ * @property originalLogins
  */
 @Serializable
 data class UserInfo(
     val name: String,
+    val id: Long? = null,
+    val oldName: String? = null,
+    val originalLogins: List<String?> = emptyList(),
```
Comment (petertrr) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:34]: Should be list of pairs, together with source?

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt [hunk 13]
```diff
@@ -19,10 +21,17 @@ import kotlinx.serialization.Serializable
...
+ * @property id
+ * @property isActive
+ * @property oldName
+ * @property originalLogins
  */
 @Serializable
 data class UserInfo(
     val name: String,
+    val id: Long? = null,
+    val oldName: String? = null,
```
Comment (petertrr) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:33]: What is `oldName`? Is it the previous login?

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Execution.kt [hunk 14]
```diff
@@ -43,7 +42,7 @@ import javax.persistence.ManyToOne
 @Entity
 class Execution(
 
-    @ManyToOne(fetch = FetchType.EAGER)
+    @ManyToOne
```
Comment (petertrr) [inline save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Execution.kt:45]: Why did you remove it? Also, isn't `EAGER` the default value?

**Review by petertrr — CHANGES_REQUESTED** [review]

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 13 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:33]: Yes. This is necessary to change the name. This is the new name and the previous one.

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 12 above)
Comment (sanyavertolet) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:34]: or even a map

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 13 above)
Comment (petertrr) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:33]: Why do we store only the previous name then? User may change the name multiple times.

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt (hunk 1 above)
Comment (petertrr) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt:131]: But why do we need to open registration page every time? We should distinguish fisrt-time users from already registered ones

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 13 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:33]: We need the old name only in the wrapper for the front, and only in order to find the user in the database by it.

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 12 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:34]: No. OriginalLogins contains both the name and the source

File: save-backend/src/test/kotlin/com/saveourtool/save/backend/controller/OrganizationControllerTest.kt (hunk 11 above)
Comment (Cheshiriks) [reply save-backend/src/test/kotlin/com/saveourtool/save/backend/controller/OrganizationControllerTest.kt]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ValidateRepository.kt (hunk 10 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ValidateRepository.kt:23]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/UsersDetailsController.kt (hunk 7 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/UsersDetailsController.kt:55]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Removed it

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt (hunk 9 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt]: This is checked further. We will open a registration page and ask the user to change their login.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: I think it's unnecessary here. The only check required is to display the response in the Input form itself.

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt (hunk 1 above)
Comment (Cheshiriks) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt:131]: Yes, on the registration view there is a check that the user is registered. And redirect the user to the project view even before rendering.

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt (hunk 9 above)
Comment (petertrr) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt]: I can see that you are trying to retrieve a user by their original username and if it's not found you start saving the user as a new user. What I'm asking is, suppose we have a user `[name = "johndoe", source = "github"]`, and then another user, `[name = "johndoe", source = "facebook"]` tries to authenticate. Seems like with current logic you would treat the new user as if they were already present, because there is a row with the same name in `original_login` table

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 5 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Fixed

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 12 above)
Comment (petertrr) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:34]: In `UserInfo` you store only names of these logins

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt (hunk 8 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt]: Fixed

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 13 above)
Comment (petertrr) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:33]: Ah, OK, so this field should be `null` except for the renaming process. Could you please add this to class KDoc?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt (hunk 9 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt]: Ok. Agree. Changed on _findByNameAndSource_

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+import kotlinx.serialization.json.Json
+
+/**
+ * `Props` retrieved from router
+ */
+external interface RegistrationProps : PropsWithChildren {
+    /**
+     * Currently logged in user or null
+     */
+    var userName: String?
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:54]: Can we pass `userInfo` from `App.kt` instead of just name?

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 13 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:33]: Added

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+                                    }
+                                    label {
+                                        input {
+                                            type = InputType.file
+                                            hidden = true
+                                            onChange = { event ->
+                                                postImageUpload(event.target)
+                                            }
+                                        }
+                                        ariaLabel = "Change organization's avatar"
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:197]:
```suggestion
                                        ariaLabel = "Change user's avatar"
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+                        className = ClassName("row justify-content-center")
+                        div {
+                            className = ClassName("col-sm-4")
+                            div {
+                                className = ClassName("container card o-hidden border-0 shadow-lg my-2 card-body p-0")
+                                div {
+                                    className = ClassName("p-5 text-center")
+                                    h1 {
+                                        className = ClassName("h4 text-gray-900 mb-4")
+                                        +"Registration new user"
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:187]:
```suggestion
                                        +"Registration of a new user"
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+                                        ariaLabel = "Change organization's avatar"
+                                        img {
+                                            className =
+                                                    ClassName("avatar avatar-user width-full border color-bg-default rounded-circle")
+                                            src = state.image?.path?.let {
+                                                "/api/$v1/avatar$it"
+                                            }
+                                                ?: run {
+                                                    "img/undraw_profile.svg"
+                                                }
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:206]:
```suggestion
                                                ?: "img/undraw_profile.svg"
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+
+    @Suppress(
+        "TOO_LONG_FUNCTION",
+        "EMPTY_BLOCK_STRUCTURE_ERROR",
+        "LongMethod",
+        "MAGIC_NUMBER",
+        "PARAMETER_NAME_IN_OUTER_LAMBDA",
+    )
+    override fun ChildrenBuilder.render() {
+        if (state.userInfo?.isActive == false) {
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:169]: Can we maybe move this check into `getUser()` in `App.kt`? I.e., when any page is opened, we get `UserInfo` and if `UserInfo.isActive == false`, redirect the user to registration page. This way we won't need to start authorization from the registration page, even if it's not visible

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt (hunk 1 above)
Comment (petertrr) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt:131]: https://github.com/saveourtool/save-cloud/pull/965#discussion_r953695679

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 19 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:177]: Moved

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 18 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 17 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 16 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 15 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Ok. Fixed

**Comment by akuleshov7** [comment]
Logic changed (bug?): Now I see my name instead of login!

**Review by akuleshov7 — CHANGES_REQUESTED** [review]
incorrect login processing

**Comment by djhdjsdkjhnkjs** [comment]
This PR does not work at all:

image

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt [hunk 20]
```diff
@@ -163,6 +163,7 @@ class App : ComponentWithScope<PropsWithChildren, AppState>() {
         HashRouter {
             requestModalHandler {
                 userInfo = state.userInfo
+
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt:166]: I tried to tinker with your PR a little and found that this solution doesn't lead to endless loop:
```kotlin
                withRouter<Props> { location, _ ->
                    if (state.userInfo?.isActive == false && !location.pathname.startsWith("/${FrontendRoutes.REGISTRATION.path}")) {
                        Navigate {
                            to = "/${FrontendRoutes.REGISTRATION.path}"
                            replace = false
                        }
                    }
                }()
```

Basically, while `Navigate` is mounted in components tree, during each render it leads to another redirect. I'm still not sure, but I think that actual `window.location` and internal location representation in react-router's state are not fully synced. In my example we are asserting on location from `HashRouter`'s state.

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/StoringServerAuthenticationSuccessHandler.kt [hunk 21]
```diff
@@ -24,6 +25,7 @@ class StoringServerAuthenticationSuccessHandler(
 ) : ServerAuthenticationSuccessHandler {
     private val logger = LoggerFactory.getLogger(javaClass)
     private val objectMapper = ObjectMapper()
+        .configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false)
```
Comment (petertrr) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/StoringServerAuthenticationSuccessHandler.kt:28]: Is it still needed? I think it causes empty arrays to be completely absent in JSONs, but how are non-empty arrays represented?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 19 above)
Comment (petertrr) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:169]: But there is still a check here

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/welcome/WelcomeView.kt [hunk 22]
```diff
@@ -113,10 +118,11 @@ class WelcomeView : AbstractView<WelcomeProps, IndexViewState>(true) {
                 }
 
                 div {
-                    Particles::class.react {
+                    // FixMe: Particles are superimposed on top of the view in some transitions
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/welcome/WelcomeView.kt:121]: Maybe more details / note that they block user interactions?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+
+    /**
+     * Image to owner avatar
+     */
+    var image: ImageInfo?
+
+    /**
+     * Currently logged in user or null
+     */
+    var userInfo: UserInfo?
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:79]: Why do we need a copy of UserInfo in component's state? Can we use value from props?

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/StoringServerAuthenticationSuccessHandler.kt (hunk 21 above)
Comment (Cheshiriks) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/StoringServerAuthenticationSuccessHandler.kt:28]: At the time of user creation, this array will always be empty. Otherwise, we will receive the Kotlin.EmptyClass class in Json and will not be able to serialize it.

Comment (petertrr) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/StoringServerAuthenticationSuccessHandler.kt:28]: Please add a comment in code about why we need this feature (btw its `EmptyList`, not `EmptyClass`) and add explicit initialization of `originalLogins` to `emptyList` in `fun Authentication.toUser()`. This way we probably won't accidentally change it to something different

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 19 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:169]: This is necessary so that you cannot enter the view by url if isActive = false and the user is redirected to the project view

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 23 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Removed it

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/StoringServerAuthenticationSuccessHandler.kt (hunk 21 above)
Comment (Cheshiriks) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/StoringServerAuthenticationSuccessHandler.kt:30]: Ok. Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 19 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Fixed

**Review by petertrr — APPROVED** [review]

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/UserDetailsService.kt [hunk 24]
```diff
@@ -22,22 +26,30 @@ import java.util.*
...
     private val userRepository: UserRepository,
+    private val originalLoginRepository: OriginalLoginRepository,
 ) : ReactiveUserDetailsService {
     private val logger = LoggerFactory.getLogger(javaClass)
 
     override fun findByUsername(username: String): Mono<UserDetails> =
-            { userRepository.findByName(username) }
-                .toMono()
-                .getIdentitySourceAwareUserDetails(username)
+            userRepository.findByName(username)?.let {
```
Comment (akuleshov7) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/UserDetailsService.kt:34]: 1) All the logic should be unified:
```Mono.just()
          .map {
                
          }
```
2) You duplicate same call twice: ` .getIdentitySourceAwareUserDetails(username)`

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 25]
```diff
@@ -0,0 +1,258 @@
...
+                                    }
+                                    ariaLabel = "Change user's avatar"
+                                    img {
+                                        className =
+                                                ClassName("avatar avatar-user width-full border color-bg-default rounded-circle")
+                                        src = state.image?.path?.let {
+                                            "/api/$v1/avatar$it"
+                                        }
+                                            ?: "img/undraw_profile.svg"
+                                        height = 260.0
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:192]: NEVER ever set width with pixels

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 26]
```diff
@@ -0,0 +1,258 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+    private fun postImageUpload(element: HTMLInputElement) =
+            scope.launch {
+                setState {
+                    isUploading = true
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:238]: Why isUploading needed here? Do you show any animation or what?

**Review by akuleshov7 — CHANGES_REQUESTED** [review]
говнокод опять

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 27]
```diff
@@ -0,0 +1,258 @@
...
+                }
+                div {
+                    className = ClassName("row justify-content-center")
+                    div {
+                        className = ClassName("col-sm-4")
+                        div {
+                            className = ClassName("container card o-hidden border-0 shadow-lg my-2 card-body p-0")
+                            div {
+                                className = ClassName("p-5 text-center")
+                                h1 {
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:172]: please split to components

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt [hunk 28]
```diff
@@ -0,0 +1,258 @@
...
+            val response = post(
+                "$apiUrl/users/save",
+                jsonHeaders,
+                Json.encodeToString(newUserInfo),
+                loadingHandler = ::classLoadingHandler,
+                responseHandler = ::classComponentResponseHandlerWithValidation,
+            )
+            if (response.ok) {
+                window.location.href = "#/${FrontendRoutes.PROJECTS.path}"
+                window.location.reload()
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:139]: why do you need to reload?

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 12 above)
Comment (akuleshov7) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:34]: you concatenated it?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 28 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt:139]: Need to update the states in App.kt otherwise the redirect will not let us leave this view

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 26 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Removed it

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt (hunk 25 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/RegistrationView.kt]: Fixed

**Review by akuleshov7 — APPROVED** [review]

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt (hunk 12 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/info/UserInfo.kt:34]: No. These are two different fields.
